core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java

```diff
@@ -0,0 +1,188 @@
+
+  BaseIncrementalChangelogScan(
+      TableOperations ops, Table table, Schema schema, TableScanContext context) {
+    super(ops, table, schema, context);
+  }
+
+  @Override
+  protected IncrementalChangelogScan newRefinedScan(
+      TableOperations newOps, Table newTable, Schema newSchema, TableScanContext newContext) {
+    return new BaseIncrementalChangelogScan(newOps, newTable, newSchema, newContext);
+  }
+
+  @Override
+  protected CloseableIterable<ChangelogScanTask> doPlanFiles(
+      Long fromSnapshotIdExclusive, long toSnapshotIdInclusive) {
+
+    Deque<Snapshot> changelogSnapshots =
+        changelogSnapshots(fromSnapshotIdExclusive, toSnapshotIdInclusive);
+
+    if (changelogSnapshots.isEmpty()) {
+      return CloseableIterable.empty();
+    }
+
+    Set<Long> changelogSnapshotIds =
+        changelogSnapshots.stream().map(Snapshot::snapshotId).collect(Collectors.toSet());
+
+    Set<ManifestFile> newDataManifests =
+        FluentIterable.from(changelogSnapshots)
+            .transformAndConcat(snapshot -> snapshot.dataManifests(table().io()))
+            .filter(manifest -> changelogSnapshotIds.contains(manifest.snapshotId()))
+            .toSet();
+
+    ManifestGroup manifestGroup =
+        new ManifestGroup(table().io(), newDataManifests, ImmutableList.of())
+            .specsById(table().specs())
+            .caseSensitive(isCaseSensitive())
+            .select(scanColumns())
+            .filterData(filter())
+            .filterManifestEntries(entry -> changelogSnapshotIds.contains(entry.snapshotId()))
+            .ignoreExisting();
+
+    if (shouldIgnoreResiduals()) {
+      manifestGroup = manifestGroup.ignoreResiduals();
+    }
+
+    if (newDataManifests.size() > 1 && shouldPlanWithExecutor()) {
+      manifestGroup = manifestGroup.planWith(planExecutor());
+    }
+
+    return manifestGroup.plan(new CreateDataFileChangeTasks(changelogSnapshots));
+  }
+
+  @Override
+  public CloseableIterable<ScanTaskGroup<ChangelogScanTask>> planTasks() {
+    return TableScanUtil.planTaskGroups(
+        planFiles(), targetSplitSize(), splitLookback(), splitOpenFileCost());
+  }
+
+  private Deque<Snapshot> changelogSnapshots(
+      Long fromSnapshotIdExclusive, long toSnapshotIdInclusive) {
+
+    Deque<Snapshot> changelogSnapshots = new ArrayDeque<>();
+
+    for (Snapshot snapshot :
+        SnapshotUtil.ancestorsBetween(table(), toSnapshotIdInclusive, fromSnapshotIdExclusive)) {
+
+      if (!snapshot.operation().equals(DataOperations.REPLACE)) {
+        if (snapshot.deleteManifests(table().io()).size() > 0) {
+          throw new UnsupportedOperationException(
+              "Delete files are currently not supported in changelog scans");
+        }
+
+        // insert at the beginning to have old snapshots first
+        changelogSnapshots.addFirst(snapshot);
+      }
+    }
+
+    return changelogSnapshots;
+  }
+
+  private static Map<Long, Integer> computeSnapshotOrdinals(Iterable<Snapshot> snapshots) {
```

Comment:
nit: should this be moved inside `CreateDataFileChangeTasks`, as it seems to be only used there?

Reply:
I am going to add another function for supporting V2 deletes. That's why I kept it outside.